File: apps/api/src/routes/analyze.ts
import { Router } from "express";
import { z } from "zod";
import type { PrismaClient } from "../generated/prisma/client.js";
import { requireAuth } from "../middleware/authMiddleware.js";
import type { AiClients } from "../services/aiClients.js";
import { createAnalyzeService } from "../services/analyzeService.js";
import { getAuthedUserId } from "./http.js";

const analyzeSchema = z.object({
  reportId: z.string().min(1),
  query: z.string().min(1),
  topK: z.number().min(1).max(20).default(5),
});

export const createAnalyzeRouter = ({
  prisma,
  vectorStore,
  aiClients,
}: {
  prisma: PrismaClient;
  vectorStore: {
    search: (queryVector: number[], reportId: string, topK: number) => Promise<{ contexts: string[]; sources: string[] }>;
  };
  aiClients: AiClients;
}) => {
  const router = Router();
  const analyzeService = createAnalyzeService({ prisma, vectorStore, aiClients });

  router.post("/", requireAuth, async (req, res) => {
    const parsed = analyzeSchema.safeParse(req.body);
    if (!parsed.success) {
      return res.status(400).json({ error: "Invalid input" });
    }

    try {
      const data = await analyzeService.answerQuestion({ userId: getAuthedUserId(req), ...parsed.data });
      if (!data.answer) {
        return res.status(404).json({ error: "Report not found" });
      }
      return res.json(data);
    } catch (error) {
      const message = error instanceof Error ? error.message : "Analysis failed";
      if (message === "Report not ready") {
        return res.status(409).json({ error: message });
      }
      return res.status(500).json({ error: message });
    }
  });

  return router;
};

File: apps/api/src/routes/analyze.py
from fastapi import APIRouter, Depends, HTTPException, Request

from src.schemas import AnalyzeRequest
from src.analyze import answer_question
from src.config import get_settings

settings = get_settings()
router = APIRouter(prefix="/api/analyze", tags=["analyze"])


def get_user_id(request: Request) -> str:
    return request.state.user_id


@router.post("")
async def analyze(
    body: AnalyzeRequest,
    request: Request,
):
    user_id = get_user_id(request)

    try:
        answer, sources = await answer_question(
            user_id=user_id,
            report_id=body.reportId,
            query=body.query,
            top_k=body.topK,
            ai_clients=request.app.state.ai_clients,
        )
        if not answer:
            raise HTTPException(status_code=404, detail="Report not found")
        return {"answer": answer, "sources": sources}
    except ValueError as e:
        if str(e) == "Report not ready":
            raise HTTPException(status_code=409, detail="Report not ready")
        raise HTTPException(status_code=500, detail=str(e))
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Analysis failed: {e}")
